trustless gateway spec: add probing path (identity cid). (#491)
ipfs/specs: src/http-gateways/trustless-gateway.md

A subset of "HTTP API" of :cite[path-gateway].

> ## `GET /ipfs/bafkqaaa`

Reviewer: Move GET/HEAD probing with identity CID to "notes for implementers" section at the end of document. Maybe call it "Inexpensive Health Check for Clients" ?

Author: I think a "dedicated probe paths" section would be enough ? See now, as it is RECOMMENDED from the description of GET/HEAD requests.

`bafkqaaa` is the identity empty CID. This endpoint can be used to probe that that the endpoint corresponds to a trustless gateway.

For block requests (`?format=raw`), when supported, it must return `200 OK` and an empty body.

> For CAR requests (`?format=car`), when supported, it must return `200 OK` and CAR file with root set to `bafkqaaa` and a single `bafkqaaa` block (which is empty).

Reviewer: go-car (used by most gateways) will skip identity blocks, which means CAR will be empty (no `Data`, just `Header`:

```suggestion
For CAR requests (`?format=car`), when supported, it must return `200 OK` and a valid, empty CAR file with CAR Header `roots` set to `bafkqaaa`.
```

Author: @lidel can you double check?

The reason I wrote it like this is because `ipfs dag export bafkqaaa > bafkqaaa.car` seemed to return a CAR with a single empty block in it (as in, https://pkg.go.dev/github.com/ipld/go-car/util#ReadNode reads a node in the car, so it doesn't seem to be just the root being set in the car header. It seems to include an empty block).

Reviewer: Yeah we have a discrepancy in behavior. I think spec should only require valid CAR with `roots` – we can't depend on `Data` because different implementations will produce different CARs. They may be logically equivalent, but won't be byte-for-byte identical.

You can inspect with `car` CLI (`go install github.com/ipld/go-car/cmd/car@latest`):

CAR produced by `boxo/gateway` ([code](https://github.com/ipfs/boxo/blob/7c3ac0d8f9e1718ebe0d637c04ab597a54a93569/gateway/backend_blocks.go#L284)) seem to correctly skip identity CID:

```console
$ curl "http://127.0.0.1:8080/ipfs/bafkqaaa?format=car" -o bafkqaaa-gw.car -s
$ car inspect bafkqaaa-gw.car
Version: 1
Roots: bafkqaaa
Root blocks present in data: No
Block count: 0
...
```

CAR produced by `ipfs dag export` ([code](https://github.com/ipfs/kubo/blob/f00e115e1ed094d47b3abe30d86fea40c6ca2f60/core/commands/dag/export.go#L24)) CLI does not not skip:

```console
$ ipfs dag export bafkqaaa > bafkqaaa-cli.car
$ car inspect bafkqaaa-cli.car
Version: 1
Roots: bafkqaaa
Root blocks present in data: Yes
Block count: 1
```

Which one is "more correct"? 
On the principle of not wasting space, identity blocks should be skipped, but they also not hurt anyone if present – just waste space because data is included twice.

Author: Ok, I take your fix.

For block requests (`?format=raw` or `Accept: application/vnd.ipld.raw`), only `GET /ipfs/{cid}[?{params}]` is supported.

> It is RECOMMENDED to additionally implement the [`GET` probe path](#dedicated-probe-paths).

Reviewer: ```suggestion
Client and Server implementations SHOULD include support for the [`GET` probe path](#dedicated-probe-paths).
```

## `HEAD /ipfs/{cid}[/{path}][?{params}]`

Same as GET, but does not return any payload.

> It is RECOMMENDED to additionally implement the [`HEAD` probe path](#dedicated-probe-paths).

Reviewer: ```suggestion
Client and Server implementations SHOULD include support for the [`HEAD` probe path](#dedicated-probe-paths).
```

Below MUST be implemented **in addition** to "HTTP Response" of :cite[path-gateway].

> ## Status codes

Reviewer: We should refer to https://specs.ipfs.tech/http-gateways/path-gateway/#response-status-codes and avoid duplicating code list. 

Follow convention from https://specs.ipfs.tech/http-gateways/trustless-gateway/#content-location-response-header and https://specs.ipfs.tech/http-gateways/trustless-gateway/#http-response where we refer to path-gateway spec where possible.

supports more than a single response format and the `format` query parameter is
missing or does not match well-known format from `Accept` header.

> ### `Retry-After` (response header)

Reviewer: Update https://specs.ipfs.tech/http-gateways/path-gateway/#retry-after-response-header and refer to that.

A Client MUST confirm the record signature match `libp2p-key` from the requested IPNS Name.

A Client MUST [perform additional record verification according to the IPNS specification](https://specs.ipfs.tech/ipns/ipns-record/#record-verification).

> # Notes for implementers

Reviewer: ```suggestion
# Appendix: Notes for implementers
```

## Dedicated Probe Paths

Trustless gateways SHOULD provide probing endpoints as described below.

### `GET /ipfs/bafkqaaa`

`bafkqaaa` is the identity empty CID. This endpoint can be used to probe that
that the endpoint corresponds to a trustless gateway.

> For block requests (`?format=raw`), when supported, it MUST return `200 OK`

Reviewer: ```suggestion
For block requests (signaled by `?format=raw` and `Accept: application/vnd.ipld.raw`), when supported, it MUST return `200 OK`
```

and an empty body.

> For CAR requests (`?format=car`), when supported, it MUST return `200 OK` and a valid CAR file with CAR Header `roots` set to `bafkqaaa`. Identity block MAY be skipped in the CAR Data section.

Reviewer: ```suggestion
For CAR requests (signaled by `?format=car` and `Accept: application/vnd.ipld.car`), when supported, it MUST return `200 OK` and a valid CAR file with CAR Header `roots` set to `bafkqaaa`. Identity block MAY be skipped in the CAR Data section.
```

A Client MUST confirm the record signature match `libp2p-key` from the requested IPNS Name.

A Client MUST [perform additional record verification according to the IPNS specification](https://specs.ipfs.tech/ipns/ipns-record/#record-verification).

# Notes for implementers

## Dedicated Probe Paths

Trustless gateways SHOULD provide probing endpoints as described below.

### `GET /ipfs/bafkqaaa`

`bafkqaaa` is the identity empty CID. This endpoint can be used to probe that
that the endpoint corresponds to a trustless gateway.

For block requests (`?format=raw`), when supported, it must return `200 OK`
and an empty body.

> For CAR requests (`?format=car`), when supported, it must return `200 OK` and
> CAR file with root set to `bafkqaaa` and a single `bafkqaaa` block (which is
> empty).

Reviewer: Given discrepancy in https://github.com/ipfs/specs/pull/491#discussion_r1962065874 this should be adjusted to require CID in header, but make data trully optional:

```suggestion
For CAR requests (`?format=car`), when supported, it MUST return `200 OK` and a valid CAR file with CAR Header `roots` set to `bafkqaaa`. Identity block MAY be skipped in the CAR Data section.
```
